org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/SignatureHelpRequestor.java

```diff
@@ -0,0 +1,211 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.contentassist;
+
+import static org.eclipse.jdt.internal.corext.template.java.SignatureUtil.getLowerBound;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.CompletionContext;
+import org.eclipse.jdt.core.CompletionProposal;
+import org.eclipse.jdt.core.CompletionRequestor;
+import org.eclipse.jdt.core.Flags;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.Signature;
+import org.eclipse.jdt.internal.corext.template.java.SignatureUtil;
+import org.eclipse.jdt.internal.corext.util.JavaModelUtil;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.handlers.CompletionResponse;
+import org.eclipse.jdt.ls.core.internal.handlers.CompletionResponses;
+import org.eclipse.jdt.ls.core.internal.javadoc.JavadocContentAccess;
+import org.eclipse.lsp4j.ParameterInformation;
+import org.eclipse.lsp4j.SignatureHelp;
+import org.eclipse.lsp4j.SignatureInformation;
+
+import com.google.common.io.CharStreams;
+import com.google.common.util.concurrent.SimpleTimeLimiter;
+import com.google.common.util.concurrent.UncheckedTimeoutException;
+
+public final class SignatureHelpRequestor extends CompletionRequestor {
+
+	private List<CompletionProposal> proposals = new ArrayList<>();
+	private final ICompilationUnit unit;
+	private CompletionProposalDescriptionProvider descriptionProvider;
+	private CompletionResponse response;
+
+	public SignatureHelpRequestor(ICompilationUnit aUnit, int offset) {
+		this.unit = aUnit;
+		response = new CompletionResponse();
+		response.setOffset(offset);
+		setRequireExtendedContext(true);
+	}
+
+	public SignatureHelp getSignatureHelp(IProgressMonitor monitor) {
+		SignatureHelp signatureHelp = new SignatureHelp();
+		response.setProposals(proposals);
+		CompletionResponses.store(response);
+
+		for (int i = 0; i < proposals.size(); i++) {
+			if (!monitor.isCanceled()) {
+				signatureHelp.getSignatures().add(this.toSignatureInformation(proposals.get(i)));
+			} else {
+				return null;

```

Comment:
returning null will likely cause some NPE or cause some invalid response from the server. Better to return an empty signaturehelp